File: 3_common_ancestors/solution.go
package main

import (
	"fmt"
)

func main() {
	tree := loadTestData()
	r := solution(tree, 10, 6)
	fmt.Println(r.Value)
}

type TreeNode struct {
	Value    int
	Children []TreeNode
}

func solution(node *TreeNode, nodeA int, nodeB int) TreeNode {
	traceA := &Trace{nodeA, false, TreeNodeStack{data: []TreeNode{}}}
	traceB := &Trace{nodeB, false, TreeNodeStack{data: []TreeNode{}}}
	computeTraces(*node, &Traces{[]*Trace{traceA, traceB}})

	for i, v := range traceA.Path.data {
		if v.Value != traceB.Path.data[i].Value {
			return traceB.Path.data[i-1]
		}
	}

	return TreeNode{}
}

type Trace struct {
	NodeValue    int
	PathComplete bool
	Path         TreeNodeStack
}

func (t *Trace) push(tn TreeNode) {
	t.Path.push(tn)
	if t.NodeValue == tn.Value {
		t.PathComplete = true
	}
}

type Traces struct {
	Data []*Trace
}

func (t *Traces) allCompleted() bool {
	for _, trace := range t.Data {
		if trace.PathComplete == false {
			return false
		}
	}
	return true
}

func (t *Traces) pushPop(pushPop bool, tn TreeNode) {
	for _, trace := range t.Data {
		if !trace.PathComplete {
			if pushPop {
				trace.push(tn)
			} else {
				trace.Path.pop()
			}
		}
	}
}

type TreeNodeStack struct {
	data []TreeNode
}

func (s *TreeNodeStack) pop() {
	s.data = s.data[:len(s.data)-1]
}

func (s *TreeNodeStack) push(tn TreeNode) {
	s.data = append(s.data, tn)
}

func computeTraces(root TreeNode, traces *Traces) {
	if traces.allCompleted() {
		return
	}

	traces.pushPop(true, root)
	if traces.allCompleted() {
		return
	}

	if len(root.Children) == 0 {
		traces.pushPop(false, TreeNode{})
	}

	for i, c := range root.Children {
		if traces.allCompleted() {
			return
		}
		computeTraces(c, traces)
		if (len(root.Children)-1) == i && !traces.allCompleted() {
			traces.pushPop(false, TreeNode{})
		}
	}
	return
}

func loadTestData() *TreeNode {
	childrenOfThree := []TreeNode{{7, []TreeNode{}}}
	childrenOfSix := []TreeNode{{8, []TreeNode{}}, {9, []TreeNode{}}}
	childrenOfFive := []TreeNode{{10, []TreeNode{}}}
	childrenOfTwo := []TreeNode{{5, childrenOfFive}, {6, childrenOfSix}}
	children := []TreeNode{{2, childrenOfTwo}, {3, childrenOfThree}, {4, []TreeNode{}}}
	return &TreeNode{1, children}
}

File: 3_common_ancestors/solution.php
<?php

class TreeNode {
    public $value;
    public $children;

    public function __construct(int $value, array $children)
    {
        $this->value = $value;
        $this->children = $children;
    }
}

class Traces {
    public $data;

    public function __construct(array $data)
    {
        $this->data = $data;
    }

    public function allComplete(): bool
    {
        foreach ($this->data as $trace) {
            if (!$trace->pathComplete) {
                return false;
            }
        }
        return true;
    }

    public function pushPop(bool $pushPop, ?TreeNode $tn)
    {
        foreach ($this->data as $trace) {
            if (!$trace->pathComplete) {
                if ($pushPop) {
                    $trace->push($tn);
                } else {
                    array_pop($trace->path);
                }
            }
        }
    }
}

class Trace {
    public $nodeValue;
    public $pathComplete = false;
    public $path = [];
    public function __construct(int $nodeValue, array $path)
    {
        $this->nodeValue = $nodeValue;
        $this->path = $path;
    }

    public function push(TreeNode $tn)
    {
        array_push($this->path, $tn);
        if ($tn->value == $this->nodeValue) {
            $this->pathComplete = true;
        }
    }
}

function loadTestData(): TreeNode
{
    $childrenOfThree = [new TreeNode(7, [])];
    $childrenOfSix = [
        new TreeNode(8, []),
        new TreeNode(9, [])
    ];
    $childrenOfFive = [new TreeNode(10, [])];
    $childrenOfTwo = [
        new TreeNode(5, $childrenOfFive),
        new TreeNode(6, $childrenOfSix)
    ];
    $children = [
        new TreeNode(2, $childrenOfTwo),
        new TreeNode(3, $childrenOfThree),
        new TreeNode(4, [])
    ];

    return new TreeNode(1, $children);
}

function main() {
    $tree = loadTestdata();
    $s = solution($tree, 8, 9);
    print_r($s->value);
}

function solution(TreeNode $rootNode, int $nodeA, int $nodeB): TreeNode {
    $traceA = new Trace($nodeA, []);
    $traceB = new Trace($nodeB, []);
    $traces = new Traces([$traceA, $traceB]);
    computeTraces($rootNode, $traces);
    foreach ($traceA->path as $i => $node) {
        if ($node->value != $traceB->path[$i]->value) {
            return $traceB->path[$i - 1];
        }
    }

    return new TreeNode(0, []);
}

function computeTraces(TreeNode $root, Traces $traces) {
    if ($traces->allComplete()) {
        return;
    }

    $traces->pushPop(true, $root);
    if ($traces->allComplete()) {
        return;
    }

    if (sizeof($root->children) == 0) {
        $traces->pushPop(false, null);
    }

    foreach ($root->children as $idx => $child) {
        if ($traces->allComplete()) {
            return;
        }
        computeTraces($child, $traces);
        if ((sizeof($root->children) - 1 == $idx) && !$traces->allComplete()) {
            $traces->pushPop(false, null);
        }
    }
}

main();